notifications_utils/template.py

```diff
@@ -109,21 +126,198 @@ def content_too_long(self):
+
+
+class WithSubjectTemplate(Template):
+
+    def __init__(
+        self,
+        template,
+        values=None
+    ):
+        self._subject = template['subject']
+        super().__init__(template, values)
 
     @property
-    def additional_data(self):
-        return self.values.keys() - self.placeholders
+    def subject(self):
+        return str(Field(self._subject, self.values))
 
-    def get_raw(self, key, default=None):
-        return self._template.get(key, default)
+    @subject.setter
+    def subject(self, value):
+        self._subject = value
 
-    def compare_to(self, new):
-        return TemplateChange(self, new)
+    @property
+    def placeholders(self):
+        return Field(self.subject).placeholders | Field(self.content).placeholders
+
+
+class PlainTextEmailTemplate(WithSubjectTemplate):
+
+    def __str__(self):
+        return Take.as_field(
+            self.content, self.values
+        ).then(
+            unlink_govuk_escaped
+        ).as_string
+
+
+class HTMLEmailTemplate(WithSubjectTemplate):
+
+    jinja_template = template_env.get_template('email_template.jinja2')
+
+    def __init__(
+        self,
+        template,
+        values=None,
+        govuk_banner=True,
+        complete_html=True,
+        brand_logo=None,
+        brand_name=None,
+        brand_colour=None
+    ):
+        super().__init__(template, values)
+        self.govuk_banner = govuk_banner
+        self.complete_html = complete_html
+        self.brand_logo = brand_logo
+        self.brand_name = brand_name
+        self.brand_colour = brand_colour and brand_colour.replace('#', '')
+
+    def __str__(self):
+
+        return self.jinja_template.render({
+            'body': get_html_email_body(
+                self.content, self.values
+            ),
+            'govuk_banner': self.govuk_banner,
+            'complete_html': self.complete_html,
+            'brand_logo': self.brand_logo,
+            'brand_name': self.brand_name,
+            'brand_colour': self.brand_colour
+        })
+
+
+class EmailPreviewTemplate(WithSubjectTemplate):
+
+    jinja_template = template_env.get_template('email_preview_template.jinja2')
+
+    def __init__(
+        self,
+        template,
+        values=None,
+        from_name=None,
+        from_address=None,
+        expanded=False,
+        show_recipient=True
+    ):
+        super().__init__(template, values)
+        self.from_name = from_name
+        self.from_address = from_address
+        self.expanded = expanded
+        self.show_recipient = show_recipient
+
+    def __str__(self):
+        return Markup(self.jinja_template.render({
+            'body': get_html_email_body(
+                self.content, self.values
+            ),
+            'subject': self.subject,
+            'from_name': self.from_name,
+            'from_address': self.from_address,
+            'recipient': Field("((email address))", template.get('values', {}), with_brackets=False),
```

Comment:
think this should be `self._template`? it's currently crashing when i try and send an email on admin app, so make sure there's a test for this